File: examples/typescript/basic.ts
import { Superserve } from "@superserve/sdk";

const client = new Superserve({ apiKey: process.env.SUPERSERVE_API_KEY });

async function main() {
  // Create a VM
  const vm = await client.vms.create({ name: "example-sandbox" });
  console.log(`Created VM: ${vm.id}`);

  // Execute a command
  const result = await client.vms.exec(vm.id, {
    command: "uname -a && echo 'Hello from Superserve!'",
  });
  console.log(`stdout: ${result.stdout}`);

  // Upload a file
  await client.files.upload(vm.id, "./hello.txt", "/home/user/hello.txt");
  console.log("Uploaded hello.txt");

  // Verify the upload
  const cat = await client.vms.exec(vm.id, { command: "cat /home/user/hello.txt" });
  console.log(`File contents: ${cat.stdout}`);

  // Checkpoint the VM
  const checkpoint = await client.checkpoints.create(vm.id, {
    name: "after-setup",
  });
  console.log(`Checkpoint created: ${checkpoint.id}`);

  // Make a change after the checkpoint
  await client.vms.exec(vm.id, { command: "rm /home/user/hello.txt" });

  // Fork from the checkpoint (hello.txt still exists in the fork)
  const forked = await client.vms.fork(vm.id, {
    checkpointId: checkpoint.id,
    name: "forked-sandbox",
  });
  console.log(`Forked VM: ${forked.id}`);

  const verify = await client.vms.exec(forked.id, {
    command: "cat /home/user/hello.txt",
  });
  console.log(`File in forked VM: ${verify.stdout}`);

  // Clean up
  await client.vms.delete(forked.id);
  await client.vms.delete(vm.id);
  console.log("Cleaned up VMs");
}

main().catch(console.error);

File: examples/python/basic.py
import os

from superserve import Superserve


def main():
    client = Superserve(api_key=os.environ["SUPERSERVE_API_KEY"])

    # Create a VM
    vm = client.vms.create(name="example-sandbox")
    print(f"Created VM: {vm.id}")

    # Execute a command
    result = client.vms.exec(vm.id, command="uname -a && echo 'Hello from Superserve!'")
    print(f"stdout: {result.stdout}")

    # Upload a file
    client.files.upload(vm.id, local_path="./hello.txt", remote_path="/home/user/hello.txt")
    print("Uploaded hello.txt")

    # Verify the upload
    cat = client.vms.exec(vm.id, command="cat /home/user/hello.txt")
    print(f"File contents: {cat.stdout}")

    # Checkpoint the VM
    checkpoint = client.checkpoints.create(vm.id, name="after-setup")
    print(f"Checkpoint created: {checkpoint.id}")

    # Make a change after the checkpoint
    client.vms.exec(vm.id, command="rm /home/user/hello.txt")

    # Fork from the checkpoint (hello.txt still exists in the fork)
    forked = client.vms.fork(vm.id, checkpoint_id=checkpoint.id, name="forked-sandbox")
    print(f"Forked VM: {forked.id}")

    verify = client.vms.exec(forked.id, command="cat /home/user/hello.txt")
    print(f"File in forked VM: {verify.stdout}")

    # Clean up
    client.vms.delete(forked.id)
    client.vms.delete(vm.id)
    print("Cleaned up VMs")


if __name__ == "__main__":
    main()
